runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamPollingScheduler.java

```diff
@@ -0,0 +1,321 @@
+ *            └────────┬──────────┴─────┘
+ *                     │ Exception
+ *                     ▼
+ *            ┌───────────────────┐
+ *            │      Error        │
+ *            └───────────────────┘
+ *
+ * On every tick the scheduler will poll/dispatch tasks only when needed, for example, when a listener is not
+ * consuming the deltas fast enough the scheduler can decide to not poll or schedule extra sync tasks until
+ * it can, this prevents superfluous polling and overwhelming the listener. Furthermore, syncing and listener
+ * notification is executed on a shared thread pool in round robin fashion, therefore if a stream produces
+ * more deltas relative to other streams that particular stream will have more CPU time than other tasks.
+ */
+
+
+@Slf4j
+public class StreamPollingScheduler {
+
+    private final Map<StreamListener, StreamingTask> allTasks = new ConcurrentHashMap<>();
+
+    private final ScheduledExecutorService scheduler;
+
+    private final ExecutorService workers;
+
+    final int pollThreshold;
+    final int pollBatchSize;
+    final Duration pollPeriod;
+
+    private final SequencerView sequencerView;
+    private final CorfuRuntime runtime;
+
+    public StreamPollingScheduler(CorfuRuntime runtime, ScheduledExecutorService scheduler, ExecutorService workers,
+                                  Duration pollPeriod, int pollBatchSize, int pollThreshold) {
+        Preconditions.checkArgument(pollBatchSize > 1, "pollBatchSize=%s has to be > 1",
+                pollBatchSize);
+        Preconditions.checkArgument(pollThreshold > 1, "pollThreshold=%s has to be > 1",
+                pollThreshold);
+        Preconditions.checkArgument(pollPeriod.toMillis() > 1, "pollPeriod=%s has to be > 1ms",
+                pollPeriod.toMillis());
+
+        this.scheduler = scheduler;
+        this.workers = workers;
+        this.pollPeriod = pollPeriod;
+        this.pollBatchSize = pollBatchSize;
+        this.pollThreshold = pollThreshold;
+        this.runtime = runtime;
+        this.sequencerView = runtime.getSequencerView();
+        tick();
+    }
+
+    public class Tick implements Runnable {
+        @Override
+        public void run() {
+            try {
+                schedule();
+            } catch (InterruptedException ie) {
+                log.warn("Scheduler thread has been interrupted!", ie);
+                Thread.currentThread().interrupt();
+            } catch (Throwable t) {
+                log.error("Unexpected throwable!", t);
+            }
+        }
+    }
+
+    private void tick() {
+        scheduler.submit(new Tick());
+    }
+
+    public void addTask(@Nonnull StreamListener streamListener, @Nonnull String namespace,
+                        @Nonnull String streamTag, @Nonnull List<String> tablesOfInterest,
+                        long lastAddress, int bufferSize) {
+
+        Preconditions.checkArgument(bufferSize >= pollThreshold);
+        synchronized (allTasks) {
```

Comment:
if you block the entire map by using synchronized, then no need for ConcurrentMap

Reply:
I thought that this synchronized was used to guard against the state of the map changing between the `containsKey` and the subsequent `put`?

Reply:
The synchronized block is not used to maintain the integrity of the `allTasks` map, its to implement a conditional variable. In Java, synchronized/notify APIs are used together to implemented a **conditional variable**.  Without a synchronized block, notify/wait semantics can lose updates, causing the thread scheduler thread to block indefinitely.

Reply:
You can read more about it here https://docs.oracle.com/javase/tutorial/essential/concurrency/guardmeth.html